---
title: "Yeast allele selection"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: hide
    collapsed: no
---


## Introduction
The purpose of this workflow is to select alleles to use 
as query alleles in query cape. It reads in allele scans
produced by setup.Rmd. These scans can be seen in 
FLZ_genes_and_QTL.html.

Identifying a good allele is kind of iterative and can
be done with multiple criteria.

To plot results from a specific allele, select a trait, 
a chromosome, an allele, and whether you want the minimum 
effect on the chrommosome or the maximum effect. This
workflow will tell you which marker to specify
in Query_CAPE.Rmd

This workflow also specifies a contrasting trait (YPD).
Ideally, you want the allele to have different effects in 
the trait of interest and in YPD. Use the contrasting plots 
at the bottom of this workflow to quickly identify chromosomes 
and alleles that have contrasting effects. Then re-set the 
parameters accordingly to identify the allele of interest.

```{r param}
trait.name = "FLZ"
chr = 15
allele = 2
min.or.max <- "min"
contrast.trait = "YPD" #a trait to contrast allele effects with
```


```{r}
library(here)
library(knitr)

all.fun <- list.files(here("Code"), full.names = TRUE, pattern = ".R")
for(i in 1:length(all.fun)){source(all.fun[i])}
load_latest_cape(here("Code", "cape"))
```

```{r read_data}
allele.effects.file <- here("Results", "General", "allele.effects.RDS") #Generated by setup.Rmd
allele.effects <- readRDS(allele.effects.file)
pheno_obj <- readRDS(here("Results", "General", "pheno_obj.RDS")) #Generated by setup.Rmd
gene.table <- read.csv(here("Data", "pgmap.csv"))
```


```{r select_query_marker}
#use this code to identify a marker and alleles to put into Query_CAPE.Rmd
trait.idx <- which(colnames(pheno_obj$pheno) == trait.name)
trait.effects <- allele.effects[[trait.idx]]
contrast.idx <- which(colnames(pheno_obj$pheno) == contrast.trait)
contrast.effects <- allele.effects[[contrast.idx]]
chr.idx <- which(pheno_obj$chromosome == chr)

if(min.or.max == "max"){
  test.idx <- which.max(trait.effects[allele,chr.idx])
}else{
  test.idx <- which.min(trait.effects[allele,chr.idx])
}

marker.name <- pheno_obj$geno_names[[3]][chr.idx[test.idx]]
```


```{r genes_near_marker}
bp.buffer = 5000
gene.chr.idx <- which(gene.table[,1] == chr)
chr.table <- gene.table[gene.chr.idx,]
marker.pos <- as.numeric(strsplit(marker.name, "_")[[1]][2])
above.min <- which(chr.table[,"start"] >= marker.pos-bp.buffer)
below.max <- which(chr.table[,"stop"] <= marker.pos+bp.buffer)
gene.idx <- intersect(above.min, below.max)
nearby.genes <- chr.table[gene.idx,]
gene.dist <- marker.pos - nearby.genes[,c("start", "stop")]
distance.to.marker <- apply(gene.dist, 1, function(x) x[which.min(abs(x))])
nearby.genes <- cbind(nearby.genes, distance.to.marker)
```

The following table shows all genes within `r bp.buffer` base pairs
of the selected marker position.

```{r nearby_gene_table}
kable(nearby.genes)
```

The marker for testing is `r marker.name`.

The following plot shows the location of this marker relative
to the allele effects of the specified allele. It also shows
this allele's effect on the contrast trait: `r contrast.trait`

It is desireable for these effects to be as distinct from one 
another as possible.

```{r plot_pos, fig.width = 9, fig.height = 4}
cols <- categorical_pal(8)
ylim <- c(min(c(trait.effects[allele,chr.idx], contrast.effects[allele,chr.idx])),
  max(c(trait.effects[allele,chr.idx], contrast.effects[allele,chr.idx])))
plot(pheno_obj$marker_location[chr.idx], trait.effects[allele,chr.idx], 
  type = "l", col = cols[allele], lwd = 2, xlab = paste("Chr", chr, "Position"),
  ylab = paste("Allele", allele, "Effect"), main = "Position of Selected Marker",
  ylim = ylim)
points(pheno_obj$marker_location[chr.idx], contrast.effects[allele,chr.idx],
  type = "l", col = cols[allele], lty = 2)
abline(v = pheno_obj$marker_location[chr.idx[test.idx]])
legend("topright", legend = c(trait.name, contrast.trait), lty = c(1,2),
  col = cols[allele], lwd = 2)
abline(h = 0)
```

## Contrast Traits {.tabset .tabset-fade .tabset-pills}

Use this part of the code to look for alleles that have dissimilar
effets in two traits of interest. For example, an interesting 
allele might be that increases growth in YPD but decreases growth
in FLZ.

The plot below shows allele effects plotted against each other
across the whole genome. 

```{r contrast_traits, fig.height = 6, fig.width = 9}

allele.cols <- categorical_pal(8)

par(mfrow = c(2,4))  
xlim = c(min(allele.effects[[trait.idx]]), max(allele.effects[[trait.idx]]))
ylim = c(min(allele.effects[[contrast.idx]]), max(allele.effects[[contrast.idx]]))
for(a in 1:nrow(allele.effects[[1]])){
 plot(allele.effects[[trait.idx]][a,], allele.effects[[contrast.idx]][a,],
    main = paste("Allele", a), xlab = trait.name, ylab = contrast.trait, 
    col = allele.cols[a], xlim = xlim, ylim = ylim)
  abline(h = 0, v = 0)
}
```

The plots below show alleles plotted against each other by 
chromosome.

```{r by_chr, fig.height = 6, fig.width = 9, results = "asis"}
xlim <- cbind(apply(allele.effects[[trait.idx]], 1, min), apply(allele.effects[[trait.idx]], 1, max))
ylim <- cbind(apply(allele.effects[[contrast.idx]], 1, min), apply(allele.effects[[contrast.idx]], 1, max))

for(ch in 1:16){
  cat("### Chr", ch, "\n")
  par(mfrow = c(2,4))  
  for(a in 1:nrow(allele.effects[[1]])){
    ch.idx <- which(pheno_obj$chromosome == ch)
    plot(allele.effects[[trait.idx]][a,ch.idx], allele.effects[[contrast.idx]][a,ch.idx],
    main = paste("Allele", a), xlab = trait.name, ylab = contrast.trait, col = allele.cols[a],
    xlim = xlim[a,], ylim = ylim[a,])
    abline(h = 0, v = 0)
  }
  cat("\n\n")
}
```