---
title: "Query CAPE"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
---

The purpose of this workflow is to explore the possiblility of
doing CAPE in a slightly different way. Instead of testing all
pairs in a region, we will test a single query marker against
all markers in the genome. 

I'm hoping that this will give us a way to start from a known
mutation and build up a network around it without getting too
crazy with the pairwise testing. 

```{r get_exp_name}
library(here)

args <- commandArgs(trailingOnly=T)
delete_previous <- as.logical(args[1])
remove_chr9 <- as.logical(args[2])
test.markers <- as.logical(args[3])
test.ind  <- as.logical(args[4])

#exp.name <- "GAL3_query"; roi <- c("GAL3" = "chr04_463411"); query.alleles <- c(1,3)
#exp.name <- "GAL4_query"; roi <- c("GAL4" = "chr16_80216"); query.alleles <- 3  
#exp.name <- "MSS11_query"; roi <- c("MSS11" = "chr13_588112"); query.alleles <- 3
#exp.name <- "GAL80_query"; roi <- c("GAL80" = "chr13_171614"); query.alleles <- 3
#exp.name <- "GAL1_7_10_query"; roi <- c("GAl1/7/10" = "chr02_271559"); query.alleles <- 6
#exp.name <- "FLZ_resistance_modifier"; roi = c("chr15" = "chr15_234651"); query.alleles = 5
exp.name <- "FLZ_susceptibility_modifier"; roi = c("chr1" = "chr1_36048"); query.alleles = 2

if(is.na(delete_previous)){
  delete_previous <- FALSE
  remove_chr9 = TRUE #removes the proximal half of Chr 9 because of allele frequency weirdness.
  test.markers <- FALSE
  test.ind <- FALSE
}

is.interactive = FALSE
#is.interactive = TRUE
results.dir <- here("Results", exp.name)
```

The results here pertain to the directory `r exp.name`.

```{r delete_old_data}
if(delete_previous){
    patterns <- c(".txt", ".csv", ".RDS", ".jpg", ".pdf")
    to.delete <- unlist(lapply(patterns, function(x) list.files(results.dir, pattern = x, full.names = TRUE)))
    unlink(to.delete)
}
```

```{r load_code}
all.fun <- list.files(here("Code"), full.names = TRUE, pattern = ".R")
for(i in 1:length(all.fun)){source(all.fun[i])}
```

```{r load_cape, echo = FALSE, warning = FALSE, message = FALSE, error = FALSE}
#load local version of cape code
needed.libraries <- c("pheatmap", "hexbin", "RColorBrewer", "gridExtra", "ggplot2")
load_libraries(needed.libraries, personal.library = TRUE)
load_latest_cape(here("Code", "cape"), personal.library = TRUE)
```

## Load data

```{r read_data}
pop.rds <- here("Data", "pop.RDS")
if(!file.exists(pop.rds)){
  pop <- read_population(filename = here("Data", "scaffold_72h_data_rqtl1_form_new_11_16_21.csv"),
    id_col = 1, na_strings = NA)
  #str(pop)
  saveRDS(pop, pop.rds)
}else{
  pop <- readRDS(pop.rds)
}

cross_obj <- cape2mpp(pop)
pheno_obj <- cross_obj[[1]]

#log transform the phenotypes
pheno_obj$pheno <- apply(pheno_obj$pheno, 2, log10)
#mean center
pheno_obj$pheno <- apply(pheno_obj$pheno, 2, function(x) x - mean(x, na.rm = TRUE))

#replace the marker_location with the actual marker_location
marker.loc <- sapply(strsplit(pheno_obj$geno_names[[3]], "_"), function(x) as.numeric(x[2]))
pheno_obj$marker_location <- marker.loc

if(remove_chr9){
    chr9.locale <- which(cross_obj[[1]]$chromosome == 9)
    to.remove <- head(chr9.locale, length(chr9.locale)/2)
    geno <- cross_obj[[2]]$geno[,,-to.remove]
    pheno_obj$chromosome <- pheno_obj$chromosome[-to.remove]
    pheno_obj$marker_num <- pheno_obj$marker_num[-to.remove]
    pheno_obj$marker_location <- pheno_obj$marker_location[-to.remove]
    pheno_obj$geno_names[[3]] <- pheno_obj$geno_names[[3]][-to.remove]
}else{
    geno <- cross_obj[[2]]$geno
}

if(test.markers){
    subsample.locale <- sort(sample(1:length(pheno_obj$chromosome), 2000))
    geno <- geno[,,subsample.locale]
    pheno_obj$chromosome <- pheno_obj$chromosome[subsample.locale]
    pheno_obj$marker_num <- pheno_obj$marker_num[subsample.locale]
    pheno_obj$marker_location <- pheno_obj$marker_location[subsample.locale]
    pheno_obj$geno_names[[3]] <- pheno_obj$geno_names[[3]][subsample.locale]
}

if(test.ind){
    subsample.locale <- sort(sample(1:nrow(pheno_obj$pheno), 2000))
    pheno_obj$pheno <- pheno_obj$pheno[subsample.locale,]
    geno <- geno[subsample.locale,,]
    pheno_obj$geno_names[[1]] <- pheno_obj$geno_names[[1]][subsample.locale]
}

#str(pheno_obj)
geno_obj <- geno
#str(geno_obj)

```


```{r variance, eval = FALSE}
#look for separation of trait variance in individuals
#with each allele at each locus. Does separation indicate
#the presence of epistasis? Main effects?

mad <- function(v){
  mean.v <- mean(v, na.rm = TRUE)
  abs.dev <- abs(v - mean.v)
  mean.dev <- mean(abs.dev, na.rm = TRUE)
  return(mean.dev)
}

for(ph in 1:ncol(pheno_obj$pheno)){
  allele.var <- matrix(NA, nrow = 8, ncol = dim(geno_obj)[3])
  overall.var <- mad(pheno_obj$pheno[,ph])
  cat("\n", colnames(pheno_obj$pheno)[ph], "\n")

  for(i in 1:dim(geno_obj)[3]){
    report.progress(i, dim(geno_obj)[3])
    allele.pheno <- lapply(1:ncol(geno_obj[,,i]), function(x) pheno_obj$pheno[which(geno_obj[,x,i] == 1),ph])
    mean.dev <- sapply(allele.pheno, function(x) mad(x))
    #dev.order <- order(mean.dev)  
    #barplot(mean.dev[dev.order], col = categorical_pal(8)[dev.order])
    #boxplot(allele.dev[dev.order], col = categorical_pal(8)[dev.order])
    allele.var[,i] <- mean.dev
    #stripchart(lapply(allele.pheno[var.order], function(x) round(x, 2)), 
    #  method = "stack", offset = 0.1, col = "lightgray",
    #  pch = 16, vertical = TRUE)
    #plot(one.var[var.order], col = categorical_pal(8)[var.order], pch = 16)
  }

  pdf(paste0("~/Desktop/variance_", colnames(pheno_obj$pheno)[ph], ".pdf"), width = 7, height = 5)
  for(ch in 1:16){
    chr.locale <- which(pheno_obj$chromosome == ch)
    chr.pos <- pheno_obj$marker_location[chr.locale]
    plot.new()
    plot.window(xlim = c(1, max(chr.pos)/1e3), ylim = c(min(allele.var, na.rm = TRUE), max(allele.var, na.rm = TRUE)))
    for(a in 1:8){
      points(x = chr.pos/1e3, allele.var[a,chr.locale], type = "l", col = categorical_pal(8)[a])
    }
    abline(h = overall.var)
    mtext(ch, side = 3)
    axis(1);axis(2)
  }
  dev.off()
}

#ph is the number or name of the phenotype you want to look at
#chr is the chromosome, and position is the position in
#base pairs to look at. The plots above are in kb, so this
#number should be multiplied by 1000
plot_one_var <- function(ph, chr, pos){
  chr.locale <- which(pheno_obj$chromosome == chr)
  chr.pos <- get.nearest.pt(pheno_obj$marker_location[chr.locale], pos)
  marker.pos <- chr.locale[chr.pos]
  allele.pheno <- lapply(1:ncol(geno_obj[,,marker.pos]), function(x) pheno_obj$pheno[which(geno_obj[,x,marker.pos] == 1),ph])
  allele.var <- sapply(allele.pheno, mad)
  var.order <- order(allele.var)
  n <- sapply(allele.pheno[var.order], length)
  stripchart(lapply(allele.pheno[var.order], function(x) round(x, 2)), 
    method = "stack", vertical = TRUE, pch = 16, offset = 0.05,
    col = categorical_pal(8)[var.order], main = paste(ph, "on Chr", chr, "at", pos/1000, "kb"),
    group.names = paste(c(1:8)[var.order], "\n", n))
  
  abline(h = median(unlist(allele.pheno), na.rm = TRUE))

}

#color by a potentially interacting locus
plot_two_var <- function(ph, chr1, pos1, chr2, pos2){
  chr1.locale <- which(pheno_obj$chromosome == chr1)
  chr1.pos <- get.nearest.pt(pheno_obj$marker_location[chr1.locale], pos1)
  marker1.pos <- chr1.locale[chr1.pos]
  
  chr2.locale <- which(pheno_obj$chromosome == chr2)
  chr2.pos <- get.nearest.pt(pheno_obj$marker_location[chr2.locale], pos2)
  marker2.pos <- chr2.locale[chr2.pos]

  marker1.allele <- lapply(1:ncol(geno_obj[,,marker1.pos]), function(x) which(geno_obj[,x,marker1.pos] == 1))
  allele.pheno <- lapply(marker1.allele, function(x) pheno_obj$pheno[x,ph])
  allele.var <- sapply(allele.pheno, mad)
  var.order <- order(allele.var)
  n <- sapply(allele.pheno[var.order], length)

  #find the genotype at marker 2 for every individual with a particular marker1 genotype
  marker2.idx <- lapply(marker1.allele, function(x) apply(geno_obj[x,,marker2.pos], 1, function(y) which(y == 1)))

  list.group <- lapply(var.order, function(x) lapply(1:8, function(y) allele.pheno[[x]][which(marker2.idx[[x]] == y)]))
  names(list.group) <- var.order
  for(i in 1:length(list.group)){
    names(list.group[[i]]) <- 1:length(list.group[[i]])
  }

  ylim = c(min(unlist(allele.pheno), na.rm = TRUE), max(unlist(allele.pheno), na.rm = TRUE))

  plot.grouped.boxes(list.group, group.cols = categorical_pal(8), 
    plot.grouping = "inner", print.vals = NA, legend.x = 56, 
    legend.y = ylim[2], main = paste0(ph, " Chr", chr1, ":", pos1, " colored by Chr", chr2, ":", pos2))


  #plot.new()
  #plot.window(xlim = c(0, 9), ylim = ylim)  
  #for(i in 1:length(allele.pheno)){
  #  x <- segment.region(i-0.2, i+0.2, length(allele.pheno))
  #  for(j in 1:length(x)){
  #    j.locale <- which(marker2.idx[[i]] == j)
  #    points(rep(x[j], length(j.locale)), allele.pheno[[var.order[i]]][j.locale], 
  #      col = categorical_pal(8)[j], pch = 16, cex = 0.5)
  #  }
  #  text(x = mean(x), y = min(ylim), labels = c(1:8)[var.order[i]])
  #}
  #axis(2)  
  #abline(h = median(unlist(allele.pheno), na.rm = TRUE))
  #mtext(paste(ph, "Chr", chr1, pos1, "colored by", "Chr", chr2, pos2), side = 3)
}


#some of the splits in variance are main effect QTLs, particularly where
#variance goes way down. 
#Some of the dips in variance seem to be just dips in allele frequency

plot_one_var(ph = "CaCl2", chr = 7, pos = 500*1000) #higher-variance alleles seem to be a bit bimodal
plot_one_var(ph = "CdCl2", chr = 2, pos = 790*1000) #overall reduced variance: just a massive QTL
plot_one_var(ph = "GAL", chr = 2, pos = 280*1000) #maybe some bimodality here too
plot_one_var(ph = "GAL", chr = 4, pos = 490*1000) #The GAL3 QTL
plot_one_var(ph = "HU", chr = 6, pos = 240*1000) #QTL, but maybe not variance QTL
plot_one_var(ph = "HU", chr = 9, pos = 390*1000) #Not sure what to make of this one
plot_one_var(ph = "HU", chr = 10, pos = 380*1000) #Dip in variance related to low n
plot_one_var(ph = "HU", chr = 15, pos = 390*1000) #Variance difference, or QTL?
plot_one_var(ph = "RAPA", chr = 3, pos = 200*1000) #bimodal sneaking in
plot_one_var(ph = "RAPA", chr = 15, pos = 500*1000) #more smeared at high end
plot_one_var(ph = "YPD", chr = 7, pos = 500*1000) #these look somewhat different

#known epistasis between alleles 1/3 and allele 6
plot_one_var(ph = "GAL", chr = 4, pos = 467*1000) #GAL3 position, main effect in 1/3 rescued by 6
plot_one_var(ph = "GAL", chr = 2, pos = 271559) #GAL1/7/10 position, rescues 1/3 effect at GAL3 locus

#look for two-way effects on the distribution
plot_two_var(ph = "GAL", chr1 = 4, pos1 = 467000, chr2 = 2, pos2 = 271559)
plot_two_var(ph = "GAL", chr1 = 4, pos1 = 467000, chr2 = 9, pos2 = 400000)
```


## Pull out single marker of interest

For query cape, we will supply a genotype vector that CAPE 
will test for interaction against all other individual alleles.

```{r get_query}
marker.pos <- sapply(strsplit(dimnames(geno_obj)[[3]], "_"), function(x) as.numeric(x[2]))
marker.idx <- rep(NA, length(roi))
names(marker.idx) <- names(roi)

for(i in 1:length(roi)){
    nearest.marker <- roi[i] 
    split.marker <- strsplit(nearest.marker, "_")
    chr <- as.numeric(gsub("chr", "", split.marker[[1]][1]))
    pos <- as.numeric(split.marker[[1]][2])
    chr.locale <- which(pheno_obj$chromosome == chr)
    pos.locale <- which(marker.pos == pos)
    marker.idx[i] <- intersect(chr.locale, pos.locale)
}

query.geno <- geno_obj[,,marker.idx]

#use one of the introgressed alleles that we know is non-functional
selected.alleles <- query.geno[,query.alleles,drop=FALSE]
query.allele <- matrix(rowSums(selected.alleles), ncol = 1)
rownames(query.allele) <- rownames(query.geno)
colnames(query.allele) <- "query-allele"
#hist(query.allele)
```

## Run CAPE

```{r run_cape}
run.finished <- as.logical(file.exists(file.path(results.dir, "Variant_Influences.csv")))
if(!run.finished){
  final_obj <- run_query_cape(pheno_obj, geno_obj,  
    query_genotype = query.allele,
    results_file = "cross.RDS", p_or_q = 1e-6, n_cores = 4, 
    initialize_only = FALSE, verbose = TRUE, run_parallel = FALSE, 
    param_file = file.path(results.dir, "cape.parameters.yml"), 
    yaml_params = NULL, results_path = results.dir, plot_pdf = TRUE)
}
```


```{r read_results}
#read in processed data
final_obj <- readRDS(file.path(results.dir, "cross.RDS"))
geno_obj <- readRDS(file.path(results.dir, "cross_geno.RDS"))
pairscan_obj <- readRDS(file.path(results.dir, "cross_pairscan.RDS"))
```

```{r var_inf, eval = FALSE}
pdf(file.path(results.dir, "Variant_Influences.pdf"), width = 8, height = 6)
plot_variant_influences_query(final_obj, geno_obj, p_or_q = 0.05)
dev.off()

```

## Null Distribution

The following code plots null distributions of cape statistics.
The plots are saved in the Results folder called `r exp.name.

```{r null_dist}
pairscan.obj <- readRDS(file.path(results.dir, "cross_pairscan.RDS"))
plot.null.dist(final_obj, pairscan.obj, path = results.dir)
```

## Variant Effects

The following code produces a pdf that shows the deviation from 
the expected trait for each interaction. The coordinates are in kb
as specified by scale_coord.

```{r var_effects}
trait.deviation.file <- file.path(results.dir, "Deviation_from_Additivity.RDS")
pdf(file.path(results.dir, "Interaction.Effects.pdf"), width = 10, height = 6)
trait_deviation <- plot_variant_effects_query(final_obj, geno_obj, 
  pheno_type = "norm_pheno", p_or_q = 1e-6, scale_coord = 1e3)
dev.off()
saveRDS(trait_deviation, trait.deviation.file)
```

The following histogram shows the distributions of 
interaction effects.

```{r inf_dist}
var.int <- write_variant_influences(final_obj, p_or_q = 1e-6, include_main_effects = FALSE)
hist(as.numeric(var.int[,"Effect"]), breaks = 100, xlab = "Interaction Effect Size",
main = "Interaction Effect Size")
#hist(as.numeric(var.int[,"|Effect|/SE"]), breaks = 100)
```

The following box plots show the trait deviation
for each trait grouped by chromosome. 

```{r deviation_by_chr, fig.width = 10, fig.height = 5}
ylim <- c(min(unlist(trait_deviation)), max(unlist(trait_deviation)))
for(ph in 1:ncol(trait_deviation[[1]][[1]])){
  if(is.interactive){quartz(width = 10, height = 5)}
  dev_by_trait <- lapply(trait_deviation, function(x) lapply(x, function(y) y[,ph]))
  plot.grouped.boxes(dev_by_trait, main = colnames(trait_deviation[[1]][[1]])[ph], 
    print.vals = NA, ylim = ylim)
  abline(h = 0)
}
```

The following figure shows a direct comparison of the deviation from additivity
for each trait. Epistasis has a larger effect on traits in the GAL environment.

```{r dev_by_trait, fig.width = 5, fig.height = 5}
dev_by_trait <- lapply(1:ncol(final_obj$pheno), function(x) lapply(trait_deviation, function(y) lapply(y, function(z) z[,x])))

boxplot(list(abs(unlist(dev_by_trait[[1]])), abs(unlist(dev_by_trait[[2]]))), 
  names = colnames(final_obj$pheno))
```

The following plots show the correlation between the deviation
from additivity for the two traits.

```{r deviation_correlation, fig.width = 10, fig.height = 6}
trait.cor <- t(sapply(trait_deviation, function(y) sapply(y, function(x) if(length(x) > 0){cor(x[,1,drop=FALSE], x[,2,drop=FALSE], use = "pairwise.complete.obs")}else{NA})))
trait.cor[which(!is.finite(trait.cor))] <- 0
barplot(trait.cor, beside = TRUE, col = c("#7fc97f", "#beaed4"))
legend("topright", fill = c("#7fc97f", "#beaed4"), legend = rownames(trait.cor))
abline(h = 0)
```

```{r check_one, eval = FALSE}

test.marker <- c("GAL3" = "X04463411")
test.marker <- c("GAl1/7/10" = "X02271559_6") #allele 6 of GAL1/7/10, should rescue GAL3 null in GAL environment
test.marker <- c("GAL4" = "X1680216") 
test.marker <- c("FLO11" = "X09393064")

pdf("~/Desktop/interactions.pdf", width = 9, height = 7)
par(mfrow = c(2,2))
test <- plot_one_variant_effect(final_obj, geno_obj, marker_name = "X12292952")
dev.off()


```

## Motif Effects

Motifs are a little different in the query context. Here I plot 
all interaction effects classified as coherent/incoherent and
aggravating/alleviating. An aggravating interaction is one that
results in a trait value more extreme than the reference genotype,
whilc an alleviating interaction is one that results in a trait value
that is less extreme than the reference genotype.

```{r motif_effects}
clin.effects <- plot_clinical_effects_query(data_obj = final_obj, geno_obj, 
    pheno_type = "norm_pheno", p_or_q = 1e-6, 
    path = results.dir, verbose = FALSE)
saveRDS(clin.effects, file.path(results.dir, "Motif_Effects.Rds"))

```

## Examples of Flipped Sign {.tabset .tabset-fade .tabset-pills}

Plots below show individual interactions that are interesting. 
We will start with the interactions for which the trait is 
affected in the opposite direction by the interaction than 
the two main effects.

```{r flipped_examples, results = "asis", fig.width = 8, fig.height = 4}
stop()
for(i in 1:length(clin.effects)){

  for(ph in 1:length(clin.effects[[i]])){
    flipped.idx <- which(clin.effects[[i]][[ph]][,"sign.flipped"] == "TRUE")
    if(length(flipped.idx) > 0){
      cat("###", names(clin.effects)[i], "{.tabset .tabset-fade .tabset-pills}\n")
      cat("####", names(clin.effects[[i]])[ph], "{.tabset .tabset-fade .tabset-pills}\n")

      flipped.table <- clin.effects[[i]][[ph]][flipped.idx,]
      flipped.allele <- sapply(strsplit(rownames(flipped.table), "_"), function(x) x[2])
      flipped.chr <- substr(rownames(flipped.table), 2,3)
      ind.examples <- unique(cbind(flipped.chr, flipped.allele))

      for(f in 1:nrow(ind.examples)){
        if(is.interactive){quartz(width = 8, height = 4)}
        par(mfrow = c(1,2))
        example.idx <- intersect(which(flipped.chr == ind.examples[f,1]), 
          which(flipped.allele == ind.examples[f,2]))
        example.table <- flipped.table[example.idx,,drop=FALSE]
        max.example.idx <- which.max(abs(as.numeric(example.table[,"actual"])))
        split.example <- strsplit(rownames(example.table)[max.example.idx], "_")
        example.marker <- split.example[[1]][1]
        example.allele <- split.example[[1]][2]
        cat("#####", paste(example.marker, example.allele, sep = "_"), "\n")
        plot_one_variant_effect(final_obj, geno_obj, 
          marker_name = example.marker, allele_name = example.allele,
          p_or_q = 1e-6, pheno_type = "norm_pheno")
        cat("\n\n")
      }
    cat("\n\n")
    cat("\n\n")
    }
  }
}

```

## Trait-Specific interactions

```{r}
int.traits <- c("FLZ")
```

The following code generates interaction plots
for markers that have a significant effect in
`r paste(int.traits, collapse = ", ")`, but 
not in the remaining traits.

It creates a pdf in the results directory called
"Trait.Specific.Interactions.pdf"

```{r plot_int}
#This function plots an interaction directly from the
#stat table so we don't have to go through plot_one_variant_effect,
#which recalculates all the effects and thus is quite slow.
allele.cols <- categorical_pal(8)
plot_int_from_stats <- function(table.row, 
  bar.labels = c("main1", "main2", "additive", "actual"), main = ""){
  effect.idx <- match(c("main1", "main2", "additive", "actual"), names(table.row))
  se.idx <- match(c("main1.se", "main2.se", "additive.se", "actual.se"), names(table.row))
  allele.idx <- grep("allele", names(table.row))
  all.effects <- as.numeric(table.row[effect.idx])
  all.se <- as.numeric(table.row[se.idx])
  plot.min <- min(all.effects - all.se)
  if(plot.min > 0){plot.min <- 0}
  plot.max <- max(all.effects + all.se)
  a <- barplot(all.effects, col = allele.cols[as.numeric(table.row[allele.idx])], 
    ylim = c(plot.min, plot.max), main = main, names = bar.labels, cex.names = 0.7)
  segments(x0 = a[,1], y0 = all.effects-all.se, y1 = all.effects+all.se)
  segments(x0 = a[,1]-0.2, x1 = a[,1]+0.2, y0 = all.effects-all.se)
  segments(x0 = a[,1]-0.2, x1 = a[,1]+0.2, y0 = all.effects+all.se)

}


```

```{r trait_spec, results = "asis", fig.width = 8, fig.height = 4}
pdf(file.path(results.dir, "Trait.Specific.Interactions.pdf"), width = 8, height = 4)
par(mfrow = c(1,2))
for(i in 1:length(clin.effects)){
  interesting.int <- vector(mode = "list", length = length(clin.effects[[i]]))
  names(interesting.int) <- names(clin.effects[[i]])
  for(ph in 1:length(clin.effects[[i]])){
      add.min <- as.numeric(clin.effects[[i]][[ph]][,"additive"]) - as.numeric(clin.effects[[i]][[ph]][,"additive.se"])
      add.max <- as.numeric(clin.effects[[i]][[ph]][,"additive"]) + as.numeric(clin.effects[[i]][[ph]][,"additive.se"])
      actual.min <- as.numeric(clin.effects[[i]][[ph]][,"actual"]) - as.numeric(clin.effects[[i]][[ph]][,"actual.se"])
      actual.max <- as.numeric(clin.effects[[i]][[ph]][,"actual"]) + as.numeric(clin.effects[[i]][[ph]][,"actual.se"])

      does.overlap <- sapply(1:length(add.min), function(x) segments.overlap(add.min[x], add.max[x], actual.min[x], actual.max[x]))
      #label each effect as interesting or not based on the specified traits above
      #if we want an interaction in one trait but not another, a significant 
      #interaction is labeled interesting in the specified trait, but it is
      #labeled interesting in other traits only if is not significant.
      if(names(clin.effects[[i]])[ph] %in% int.traits){
        is.interesting <- !does.overlap
      }else{
        is.interesting <- does.overlap
      }
      interesting.int[[ph]] <- clin.effects[[i]][[ph]][which(is.interesting),]
  }
    #str(interesting.int)
    trait.spec <- Reduce("intersect", lapply(interesting.int, rownames))

    if(length(trait.spec) > 0){
      par(mfrow = c(1,2))
      for(f in 1:length(trait.spec)){
        spec.marker <- trait.spec[f]
        marker.idx <- lapply(interesting.int, function(x) which(rownames(x) == spec.marker))
        
        if(is.interactive){quartz(width = 8, height = 4); par(mfrow = c(1,2))}
          plot.chr <- substr(trait.spec[f], 2, 3)
          split.marker <- strsplit(trait.spec[f], "_")[[1]]
          marker.allele <- split.marker[[2]]
          marker.pos <- substr(split.marker[1], 4, nchar(split.marker[1]))
          bar.labels <- list(c("query", trait.spec[f], "additive", "actual"),
            c(trait.spec[f], "query", "additive", "actual"))
        
          null <- lapply(1:length(interesting.int), 
            function(x) plot_int_from_stats(interesting.int[[x]][marker.idx[[x]],],
            bar.labels = bar.labels[[x]], names(interesting.int)[x]))
          plot.label <- paste0("Chr ", plot.chr, " at ", marker.pos, "; allele ", marker.allele)
          mtext(plot.label, side = 3, outer = TRUE, line = -2, font = 2)
      }
    }
    all.int.effects <- lapply(interesting.int, 
      function(x) apply(x[trait.spec,c("main1", "main2", "additive", "actual")], 2, as.numeric))
    test <- Reduce("rbind", all.int.effects)
    test <- Reduce("cbind", all.int.effects)
    pheatmap(test, cluster_cols = FALSE)
    plot.decomp(test, cols = allele.cols[alleles[[1]]])
    add.diff <- lapply(all.int.effects, function(x) x[,"actual"] - x[,"additive"])

    test <- Reduce("cbind", add.diff)
    colnames(test) <- names(add.diff)
    #pheatmap(test, cluster_cols = FALSE)
    row.order <- hclust(dist(test))$order
    pdf("~/Desktop/test.pdf", width = 7, height = 15)
    imageWithText(test[row.order,], show.text = FALSE, split.at.vals = TRUE, 
      col.scale = c("blue", "brown"), grad.dir = "ends", col.text.rotation = 0,
      col.text.adj = 0.5, col.text.shift = 0.05, row.names = trait.spec[row.order],
      row.text.shift = 0.1)
    dev.off()
    alleles <- lapply(interesting.int, function(x) as.numeric(x[trait.spec,12]))
    par(mfrow = c(2,3))
    for(x in 1:ncol(all.int.effects[[1]])){
      if(i == 1){
        plot.labels <- c("query", "test", "additive", "actual")
      }else{
        plot.labels <- c("test", "query", "additive", "actual")
      }
      plot.lim <- c(min(c(all.int.effects[[1]][,x], all.int.effects[[2]][,x])), 
        max(c(all.int.effects[[1]][,x], all.int.effects[[2]][,x])))
      plot(all.int.effects[[1]][,x], all.int.effects[[2]][,x], xlim = plot.lim,
      ylim = plot.lim, col = allele.cols[alleles[[1]]], pch = 16,
      xlab = names(all.int.effects)[1], ylab = names(all.int.effects)[2],
      main = plot.labels[x])
      abline(h = 0, v = 0)
    }
    plot(add.diff[[1]], add.diff[[2]], xlab = names(all.int.effects)[1], 
      ylab = names(all.int.effects)[2], main = "Deviation from Additive",
      pch = 16, col = allele.cols[alleles[[1]]])
    abline(h = 0, v = 0)
  }
dev.off()

```


```{r synthetic_viability, eval = FALSE}
#look for allele pairs that are never found together
#or never separated from the query.

query_genotype <- final_obj$query_genotype

query_genotype <- geno_obj[,sample(1:8, 1), sample(1:dim(geno_obj)[3], 1)] #random query

query_cor <- matrix(NA, nrow = 8, ncol = dim(geno_obj)[3])
for(i in 1:dim(geno_obj)[3]){
  report.progress(i, dim(geno_obj)[3])
  query_cor[,i] <- apply(geno_obj[,,i], 2, function(x) cor(query_genotype, x, use = "pairwise.complete.obs"))
}


plot.new()
plot.window(xlim = c(1, dim(geno_obj)[3]), ylim = c(-1, 1))
for(i in 1:nrow(query_cor)){
  points(query_cor[i,], type = "l", col = categorical_pal(8)[i])
}
axis(2)

all.pairs <- pair.matrix(1:dim(geno_obj)[3])
full.cor <- matrix(NA, nrow = dim(geno_obj)[3], ncol = dim(geno_obj)[3])
for(i in 1:nrow(all.pairs)){
  marker1 <- all.pairs[i,1]
  marker2 <- all.pairs[i,2]
  full.cor[marker1, marker2] <- cor(as.vector(geno_obj[,,marker1]), as.vector(geno_obj[,,marker2]), use = "pairwise.complete.obs")
}

```
