---
title: "Yeast allele selection"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
---


## Introduction
The purpose of this workflow is to select alleles to use 
as query alleles in query cape. It reads in allele scans
produced by setup.Rmd. These scans can be seen in 
FLZ_genes_and_QTL.html.

To identify a good allele, pick a trait, a chromosome 
an allele, and whether you want the minimum effect
on the chrommosome or the maximum effect. This
workflow will tell you which marker to specify
in Query_CAPE.Rmd

```{r param}
trait.name = "FLZ"
chr = 13
allele = 3
min.or.max <- "max"
```


```{r}
library(here)

all.fun <- list.files(here("Code"), full.names = TRUE, pattern = ".R")
for(i in 1:length(all.fun)){source(all.fun[i])}
load_latest_cape(here("Code", "cape"))
```

```{r read_data}
allele.effects.file <- here("Results", "General", "allele.effects.RDS") #Generated by setup.Rmd
allele.effects <- readRDS(allele.effects.file)
pheno_obj <- readRDS(here("Results", "General", "pheno_obj.RDS")) #Generated by setup.Rmd
#geno_obj <- readRDS(here("Results", "General", "geno_obj.RDS")) #Generated by setup.Rmd
```


```{r select_query_marker, eval = FALSE}
#use this code to identify a marker and alleles to put into Query_CAPE.Rmd
trait.idx <- which(colnames(pheno_obj$pheno) == trait.name)
trait.effects <- allele.effects[[trait.idx]]
chr.idx <- which(pheno_obj$chromosome == chr)

if(max.or.min == "max"){
  test.idx <- which.max(trait.effects[allele,chr.idx])
}else{
  test.idx <- which.min(trait.effects[allele,chr.idx])
}

marker.name <- pheno_obj$geno_names[[3]][chr.idx[test.idx]]

cols <- categorical_pal(8)
plot(pheno_obj$marker_location[chr.idx], trait.effects[allele,chr.idx], 
  type = "l", col = cols[allele], lwd = 2, xlab = paste("Chr", chr, "Position"),
  ylab = paste("Allele", allele, "Effect"), main = "Position of Selected Marker")
abline(v = pheno_obj$marker_location[chr.idx[test.idx]])
```