---
title: "Query CAPE"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
---

The purpose of this workflow is to explore the possiblility of
doing CAPE in a slightly different way. Instead of testing all
pairs in a region, we will test a single query marker against
all markers in the genome. 

I'm hoping that this will give us a way to start from a known
mutation and build up a network around it without getting too
crazy with the pairwise testing. 

```{r get_exp_name}
library(here)

args <- commandArgs(trailingOnly=T)
exp.name = args[1]
delete_previous <- as.logical(args[2])
remove_chr9 <- as.logical(args[3])
testing <- as.logical(args[4])

if(is.na(exp.name)){
  exp.name <- "GAL_query"
  delete_previous <- TRUE
  remove_chr9 = TRUE #removes the proximal half of Chr 9 because of allele frequency weirdness.
  testing <- TRUE
}

roi <- c("GAL3" = "chr04_463411")
query.alleles <- c(1,3)

results.dir <- here("Results", exp.name)
```

The results here pertain to the directory `r exp.name`.

```{r delete_old_data}
if(delete_previous){
    patterns <- c(".txt", ".csv", ".RDS", ".jpg", ".pdf")
    to.delete <- unlist(lapply(patterns, function(x) list.files(results.dir, pattern = x, full.names = TRUE)))
    unlink(to.delete)
}
```

```{r load_code}
all.fun <- list.files(here("Code"), full.names = TRUE, pattern = ".R")
for(i in 1:length(all.fun)){source(all.fun[i])}
```

```{r load_cape, echo = FALSE, warning = FALSE, message = FALSE, error = FALSE}
#load local version of cape code
needed.libraries <- c("pheatmap")
load_libraries(needed.libraries, personal.library = TRUE)
load_latest_cape(here("Code", "cape"), personal.library = TRUE)
```

## Load data

```{r read_data}
pop.rds <- here("Data", "pop.RDS")
if(!file.exists(pop.rds)){
  pop <- read_population(filename = here("Data", "scaffold_72h_data_rqtl1_form_new_11_16_21.csv"),
    id_col = 1, na_strings = NA)
  #str(pop)
  saveRDS(pop, pop.rds)
}else{
  pop <- readRDS(pop.rds)
}

cross_obj <- cape2mpp(pop)
pheno_obj <- cross_obj[[1]]

#log transform the phenotypes
pheno_obj$pheno <- apply(pheno_obj$pheno, 2, log10)
#mean center
pheno_obj$pheno <- apply(pheno_obj$pheno, 2, function(x) x - mean(x, na.rm = TRUE))

if(remove_chr9){
    chr9.locale <- which(cross_obj[[1]]$chromosome == 9)
    to.remove <- head(chr9.locale, length(chr9.locale)/2)
    geno <- cross_obj[[2]]$geno[,,-to.remove]
    pheno_obj$chromosome <- pheno_obj$chromosome[-to.remove]
    pheno_obj$marker_num <- pheno_obj$marker_num[-to.remove]
    pheno_obj$marker_location <- pheno_obj$marker_location[-to.remove]
    pheno_obj$geno_names[[3]] <- pheno_obj$geno_names[[3]][-to.remove]
}else{
    geno <- cross_obj[[2]]$geno
}

if(testing){
    subsample.locale <- sort(sample(1:length(pheno_obj$chromosome), 2000))
    geno <- geno[,,subsample.locale]
    pheno_obj$chromosome <- pheno_obj$chromosome[subsample.locale]
    pheno_obj$marker_num <- pheno_obj$marker_num[subsample.locale]
    pheno_obj$marker_location <- pheno_obj$marker_location[subsample.locale]
    pheno_obj$geno_names[[3]] <- pheno_obj$geno_names[[3]][subsample.locale]
}

#str(pheno_obj)
geno_obj <- geno
#str(geno_obj)

```

## Pull out single marker of interest

For query cape, we will supply a genotype vector that CAPE 
will test for interaction against all other individual alleles.

```{r gal3}

marker.pos <- sapply(strsplit(dimnames(pop$geno)[[3]], "_"), function(x) as.numeric(x[2]))
marker.idx <- rep(NA, length(roi))
names(marker.idx) <- names(roi)

for(i in 1:length(roi)){
    nearest.marker <- roi[i] 
    split.marker <- strsplit(nearest.marker, "_")
    chr <- as.numeric(gsub("chr", "", split.marker[[1]][1]))
    pos <- as.numeric(split.marker[[1]][2])
    chr.locale <- which(pop$chromosome == chr)
    pos.locale <- which(marker.pos == pos)
    marker.idx[i] <- intersect(chr.locale, pos.locale)
}

gal.geno <- pop$geno[,,marker.idx]

#use one of the introgressed alleles that we know is non-functional
selected.alleles <- gal.geno[,query.alleles,drop=FALSE]
gal.null <- matrix(rowSums(selected.alleles), ncol = 1)
rownames(gal.null) <- rownames(gal.geno)
colnames(gal.null) <- "GAL-query"
#hist(gal.null)
```

## Run CAPE

```{r run_cape}
run.finished <- as.logical(file.exists(file.path(results.dir, "Variant_Influences.csv")))

if(!run.finished){
  final_obj <- run_query_cape(pheno_obj, geno_obj,  
    query_genotype = gal.null,
    results_file = "cross.RDS", p_or_q = 1e-6, n_cores = 4, 
    initialize_only = FALSE, verbose = TRUE, run_parallel = FALSE, 
    param_file = file.path(results.dir, "cape.parameters.yml"), 
    yaml_params = NULL, results_path = results.dir, plot_pdf = TRUE)
}else{
  final_obj <- readRDS(file.path(results.dir, "cross.RDS"))
  geno_obj <- readRDS(file.path(results.dir, "cross_geno.RDS"))
  pairscan_obj <- readRDS(file.path(results.dir, "cross_pairscan.RDS"))
}
```

## Null Distribution

The following code plots null distributions of cape statistics.
The plots are saved in the Results folder called `r exp.name.

```{r null_dist}
pairscan.obj <- readRDS(file.path(results.dir, "cross_pairscan.RDS"))
plot.null.dist(final_obj, pairscan.obj, path = results.dir)
```
