---
title: "Yeast Run CAPE"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
---

The purpose of this workflow is to run cape on the yeast data.
The experiment name specifies which parameter file to use.

```{r get_exp_name}
library(here)

args <- commandArgs(trailingOnly=T)
exp.name = args[1]
delete_previous <- as.logical(args[2])

if(is.na(exp.name)){
    exp.name <- "chr15"
    delete_previous <- FALSE
}

results.dir <- here("Results", exp.name)
remove_chr9 = TRUE
subsample.markers = TRUE
```

The results here pertain to the directory `r exp.name`.

```{r delete_old_data}
if(delete_previous){
    patterns <- c(".txt", ".csv", ".RDS", ".jpg", ".pdf")
    to.delete <- unlist(lapply(patterns, function(x) list.files(results.dir, pattern = x, full.names = TRUE)))
    unlink(to.delete)
}
```

```{r load_code}
all.fun <- list.files(here("Code"), full.names = TRUE, pattern = ".R")
for(i in 1:length(all.fun)){source(all.fun[i])}
```

```{r load_cape, echo = FALSE, warning = FALSE, message = FALSE, error = FALSE}
#load local version of cape code
needed.libraries <- c("pheatmap")
load_libraries(needed.libraries, personal.library = TRUE)
load_latest_cape(here("Code", "cape"), personal.library = TRUE)
```

Load file and run cape.

```{r read_data}
pop <- read_population(filename = here("Data", "scaffold_72h_data_rqtl1_form_new_11_16_21.csv"),
  id_col = 1, na_strings = NA)
#str(pop)

cross_obj <- cape2mpp(pop)
pheno_obj <- cross_obj[[1]]

#log transform the phenotypes
pheno_obj$pheno <- apply(pheno_obj$pheno, 2, log10)
#mean center
pheno_obj$pheno <- apply(pheno_obj$pheno, 2, function(x) x - mean(x, na.rm = TRUE))

if(remove_chr9){
  geno <- cross_obj[[2]]$geno[,,which(cross_obj[[1]]$chromosome != 9)]
}else{
  geno <- cross_obj[[2]]$geno
}

#str(pheno_obj)
geno_obj <- geno
#str(geno_obj)

```

If specified, subsample the genome to regions 
identified by fine-mapping. This will help 
speed up the cape run.

```{r subsample}
marker.pos <- sapply(strsplit(dimnames(geno_obj)[[3]], "_"), function(x) as.numeric(x[2]))

if(subsample.markers){

  get_region_idx <- function(region.chr, region.start, region.end, all.chr, all.pos){
    chr.idx <- which(all.chr == region.chr)
    greater.than.min <- which(all.pos[chr.idx] >= region.start)
    less.than.max <- which(all.pos[chr.idx] <= region.end)
    region.idx <- intersect(greater.than.min, less.than.max)
    if(length(region.idx) == 0){
      region.idx <- c(chr.idx[tail(less.than.max, 1)], chr.idx[head(greater.than.min, 1)])
    }
    return(chr.idx[region.idx])
  }

  marker.chr <- pheno_obj$chromosome

  param <- read_parameters(here("Results", exp.name, "cape.parameters.yml"))
  traits <- param$traits

  unique.marker.idx <- vector(mode = "list", length = length(traits))
  names(unique.marker.idx) <- names(traits)

  for(tr in 1:length(traits)){
    fine.map <- read.delim(here("Data", "Fine_mapping_for_Anna", paste0("72h_", traits[tr], "_aov_type2_genes_filtered_v9.txt")))
    region.chrs <- as.numeric(gsub("chr", "", fine.map[,"Chrom"]))
    left.lod.pos <- fine.map[,"Left"]
    right.lod.pos <- fine.map[,"Right"]

    all.region.markers <- sapply(1:length(region.chrs), 
      function(x) get_region_idx(region.chr = region.chrs[x], 
      region.start = left.lod.pos[x], region.end = right.lod.pos[x], 
      all.chr = marker.chr, all.pos = marker.pos))

    unique.marker.idx[[tr]] <- sort(unique(unlist(all.region.markers)))

    #length(unique(unlist(unique.marker.idx[[tr]])))
  }

  unique.markers <- sort(unique(unlist(unique.marker.idx)))
  geno_obj <- geno_obj[,,unique.markers]
  pheno_obj$geno_names[[3]] <- pheno_obj$geno_names[[3]][unique.markers]
  pheno_obj$chromosome <- pheno_obj$chromosome[unique.markers]
  pheno_obj$marker_num <- pheno_obj$marker_num[unique.markers]
  pheno_obj$marker_location <- marker.pos[unique.markers]
}else{
  #for some reason my conversion function is not giving us good 
  #marker positions. Replace them with the real marker positions.
  pheno_obj$marker_location <- marker.pos
}

```

```{r run_cape}
run.finished <- as.logical(file.exists(file.path(results.dir, "Variant_Influences.csv")))

if(!run.finished){
  final_obj <- run_cape(pheno_obj, geno_obj, 
    results_file = "cross.RDS", p_or_q = 0.05, n_cores = 4, 
    initialize_only = FALSE, verbose = TRUE, run_parallel = FALSE, 
    param_file = file.path(results.dir, "cape.parameters.yml"), 
    yaml_params = NULL, results_path = results.dir)
}else{
  final_obj <- readRDS(file.path(results.dir, "cross.RDS"))
  geno_obj <- readRDS(file.path(results.dir, "cross_geno.RDS"))
  pairscan_obj <- readRDS(file.path(results.dir, "cross_pairscan.RDS"))
}
```

## Null Distribution

The following code plots null distributions of cape statistics.
The plots are saved in the Results folder called `r exp.name.

```{r null_dist}
pairscan.obj <- readRDS(file.path(results.dir, "cross_pairscan.RDS"))
plot.null.dist(final_obj, pairscan.obj, path = results.dir)
```

## Motifs

Look at network motifs.

```{r motif_enrichment, fig.width = 7, fig.height = 5}
motif.file <- here(results.dir, "Motif.Results.RDS")

if(!file.exists(motif.file)){
  motif.obj <- find.motifs(final_obj)
  motif.count <- count.motifs(motif.obj)
  motif.enrich <- motif.enrichment(final_obj)
  saveRDS(motif.enrich, motif.file)
}else{
  motif.enrich <- readRDS(motif.file)
}

#jpeg("~/Desktop/motifs.jpg", width = 7, height = 5, units = "in", res = 300)
fig.cols <- c(1:3, 6:8)
par(mar = c(8, 4, 4, 2))
main.shift = 0.3; col.text.shift = 0.27; row.text.shift = 0.3; col.text.rotation = 45
imageWithText(motif.enrich$overlayed[,fig.cols], 
    col.names = colnames(motif.enrich$motif.counts)[fig.cols], 
			row.names = rownames(motif.enrich$motif.counts), 
      main = "Overlay of Enriched (brown) and Depleted (blue) Motifs", 
			class.mat = motif.enrich$overlayed.class.mat[,fig.cols], 
      grad.dir = "low", color.fun = "exponential", exp.steepness = 10, 
			global.color.scale = FALSE, global.min = 0, global.max = 1, cex = 1.2, 
			col.text.adj = 1, main.shift = main.shift, col.text.shift = col.text.shift, 
			row.text.shift = row.text.shift, col.text.rotation = col.text.rotation,
      col.scale = c("brown", "blue"), light.dark = "l")	 	
#dev.off()
```

## Allele Interaction Patterns

The following code focuses on strain effects both 
for main effects and interactions.

```{r strain_main_effects}
all.inf <- write_variant_influences(final_obj, p_or_q = 0.05, write_file = FALSE)
main.locale <- which(all.inf[,"Target"] %in% colnames(final_obj$pheno))
just.main <- all.inf[main.locale,]
main.marker <- just.main[,"Source"]
split.marker <- strsplit(main.marker, "_")
just.marker <- sapply(split.marker, function(x) x[1])
just.allele <- sapply(split.marker, function(x) x[2])
```

The following bar plot shows the number of main effects derived
from each strain.

```{r plot_num}
barplot(table(just.allele), ylab = "Number of Main Effects")
```

The following boxplot shows the sign and magnitude of these
main effects for each trait.

```{r box_main, fig.width = 8, fig.height = 5}
u_alleles = sort(unique(just.allele))
u_pheno <- colnames(final_obj$pheno)
pheno_effects <- vector(mode = "list", length = length(u_pheno))
names(pheno_effects) <- u_pheno
for(i in 1:length(u_pheno)){
  allele_effects <- vector(mode = "list", length = length(u_alleles))
  names(allele_effects) <- u_alleles
  pheno.locale <- which(just.main[,"Target"] == u_pheno[i])
  for(j in 1:length(u_alleles)){
    allele.locale <- which(just.allele == u_alleles[j])
    overlap.locale <- intersect(allele.locale, pheno.locale)
    allele_effects[[j]] <- as.numeric(just.main[overlap.locale,"Effect"])
  }
  pheno_effects[[i]] <- allele_effects
}
plot.grouped.boxes(pheno_effects, legend.x = 17, legend.y = -0.1)
abline(h = 0)
```

The following heat map shows how many interactions were 
identified between each of the allele pairs.

```{r allele_int}
just.int <- all.inf[which(is.na(all.inf[,"conditioning_marker"])),]
source.marker <- just.int[,"Source"]
target.marker <- just.int[,"Target"]
split.source <- strsplit(source.marker, "_")
source.allele <- sapply(split.source, function(x) x[2])
split.target <- strsplit(target.marker, "_")
target.allele <- sapply(split.target, function(x) x[2])

allele.int <- cbind(source.allele, target.allele)
int.count <- as.matrix(table(data.frame(allele.int)))
imageWithText(int.count, col.scale = "blue", row.text.shift = 0.1,
col.text.shift = 0.12, col.text.rotation = 0, cex = 1.5)

#int.net <- graph_from_edgelist(allele.int)
#plot(int.net)
```

```{r combo_effects}
source.effect <- vector(mode = "list", length = length(u_alleles))
names(source.effect) <- u_alleles
mean.effect <- matrix(NA, nrow = 8, ncol = 8)
rownames(mean.effect) <- colnames(mean.effect) <- 1:8
for(sallele in u_alleles){
  target.effect <- vector(mode = "list", length = length(u_alleles))
  names(target.effect) <- u_alleles
  source.locale <- which(source.allele == sallele)
  for(tallele in u_alleles){
    target.locale <- which(target.allele == tallele)
    overlap.locale <- intersect(source.locale, target.locale)
    target.effect[[tallele]] <- abs(as.numeric(just.int[overlap.locale,"Effect"]))
    mean.effect[sallele,tallele] <- median(as.numeric(just.int[overlap.locale,"Effect"]))
  }
  source.effect[[sallele]] <- target.effect
}
```

The following boxplot shows the distribution of effects 
of interactions between each pair of alleles. The effects
are grouped first by source allele then by target allele.

```{r plot_boxes}
plot.grouped.boxes(source.effect)
abline(h = 0)
```

The following heat map shows the median effect across 
all interacting pairs of alleles. Interactions within
strain are along the diagonal.

```{r plot_mat}
imageWithText(round(mean.effect, 2), col.scale = c("blue", "brown"), 
split.at.vals = TRUE, row.text.shift = 0.1, grad.dir = "ends",
col.text.shift = 0.12, col.text.rotation = 0, cex = 1.5)
```

The following boxes show the magnitude of effects of interactions
within strain and between strain.

```{r agg_plot}
boxplot(list(abs(diag(mean.effect)), c(abs(mean.effect[upper.tri(mean.effect)]), 
  abs(mean.effect[lower.tri(mean.effect)]))), names = c("Within", "Between"))
```

The only pattern I can see is that alleles with more main effects
also have more interactions.

## Motif Effects {.tabset .tabset-fade .tabset-pills}

```{r motif_effects, fig.width = 10, fig.height = 8}
#pdf("~/Desktop/effect.pdf", width = 10, height = 8)
pheno.effect.mats <- pheno.effects.DO(data.obj = final_obj, 
  geno.obj = geno_obj, color.scheme = "other", geno.coding = "Additive")
#dev.off()
```

```{r ind_effects, eval = FALSE}
int.type = 2
hist(as.numeric(pheno.effect.mats[[int.type]][,"Actual"]))
idx <- which(as.numeric(pheno.effect.mats[[int.type]][,"Actual"]) < -0.15)
#idx <- 14
test <- plot_effects(final_obj, geno_obj, marker1 = pheno.effect.mats[[int.type]][idx,"Source"], 
  marker2 = pheno.effect.mats[[int.type]][idx,"Target"], plot_type = "b", error_bars = "se",
  gen_model1 = "Additive", gen_model2 = "Additive")

```

In the following boxplots, values greater than 0 indicate a stabilizing 
interaction and values less than 0 indicate a destabilizing interaction.

The bar plots show proportions of each type of interaction for between-strain
interactions and within-strain interactions.

```{r effects_of_ints, fig.height = 8, fig.width = 4, results = "asis"}
classes <- c("more_than_additive", "less_than_additive", "extreme")
for(i in 1:length(pheno.effect.mats)){
  cat("###", names(pheno.effect.mats)[i], "\n")
  split.source <- strsplit(pheno.effect.mats[[i]][,"Source"], "_")
  source.allele <- sapply(split.source, function(x) x[2])
  split.target <- strsplit(pheno.effect.mats[[i]][,"Target"], "_")
  target.allele <- sapply(split.target, function(x) x[2])
  source.target.int <- cbind(source.allele, target.allele)
  within.idx <- which(apply(source.target.int, 1, function(x) x[1] == x[2]))
  between.idx <- which(apply(source.target.int, 1, function(x) x[1] != x[2]))
  within.class <- pheno.effect.mats[[i]][within.idx,"Class"]
  between.class <- pheno.effect.mats[[i]][between.idx,"Class"]
  test.bet <- as.numeric(pheno.effect.mats[[i]][between.idx,"Actual"])-as.numeric(pheno.effect.mats[[i]][between.idx,"Additive"])
  test.with <- as.numeric(pheno.effect.mats[[i]][within.idx,"Actual"])-as.numeric(pheno.effect.mats[[i]][within.idx,"Additive"])
  
  #in the following plot values greater than 0 indicate a stabilizing 
  #interaction and values less than 0 indicate a destabilizing interaction
  #quartz(width = 4, height = 8)
  par(mar = c(4,4,4,4))
  layout(matrix(c(1,2,3), ncol = 1), heights = c(1, 1, 0.3))
  boxplot(list(test.bet, test.with), names = c("Between", "Within"))
  stripchart(list(test.bet, test.with), method = "jitter", vertical = TRUE, add = TRUE)
  abline(h = 0)

  within.class.counts <- sapply(classes, function(x) length(which(within.class == x)))
  within.prop <- within.class.counts/sum(within.class.counts)
  between.class.counts <- sapply(classes, function(x) length(which(between.class == x)))
  between.prop <- between.class.counts/sum(between.class.counts)
  prop.table <- rbind(within.prop, between.prop)
  
  class.col = c("#dfc27d", "#80cdc1", "#d7191c")
  barplot(t(prop.table), beside = FALSE, main = names(pheno.effect.mats)[i], 
    col = class.col, names = c("Within-Strain", "Between_Strain"))
  
  par(mar = c(2,0,0,0))
  plot.new()
  plot.window(xlim = c(0,1), ylim = c(0,1))
  x.vals <- segment.region(0,0.75, 3, "ends")
  y.vals = rep(1, length(x.vals))
  rectangle.height = 0.3
  par(xpd = NA)
  nulls <- sapply(1:length(x.vals), function(x) 
    draw.rectangle(x.vals[x], x.vals[x]+0.05, y.vals[x], y.vals[x]-rectangle.height, 
    fill = class.col[x]))
  nulls <- sapply(1:length(x.vals), function(x) 
    text(x = x.vals[x]+0.06, y = (y.vals[x]-(rectangle.height/2)), labels = colnames(prop.table)[x], 
    adj = 0, cex = 1.2))
  par(xpd = TRUE)
  cat("\n\n")
}


```
