---
title: "Yeast CAPE Setup"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
---


## Introduction

The purpose of this workflow is to start looking at the yeast data from 
Aimee Dudley and Gareth Cromie. We look at allele frequencies, trait
correlations, and mapping of individual traits to help identify good
candidates for running CAPE.


```{r load_code}
library(here)
all.fun <- list.files(here("Code"), full.names = TRUE, pattern = ".R")
for(i in 1:length(all.fun)){source(all.fun[i])}
```

```{r load_cape, echo = FALSE, warning = FALSE, message = FALSE, error = FALSE}
#load local version of cape code
needed.libraries <- c("pheatmap")
load_libraries(needed.libraries, personal.library = TRUE)
load_latest_cape(here("Code", "cape"), personal.library = TRUE)
```


Load file and run cape.

```{r read_data}
pop <- read_population(filename = here("Data", "scaffold_72h_data_rqtl1_form_new_11_16_21.csv"),
  id_col = 1, na_strings = NA)
#str(pop)

cross_obj <- cape2mpp(pop)
pheno_obj <- cross_obj[[1]]

#log transform the phenotypes
pheno_obj$pheno <- apply(pheno_obj$pheno, 2, log10)
#mean center
pheno_obj$pheno <- apply(pheno_obj$pheno, 2, function(x) x - mean(x, na.rm = TRUE))

#str(pheno_obj)
geno_obj <- cross_obj[[2]]$geno
#str(geno_obj)

#fix marker position.
marker.pos <- as.numeric(sapply(strsplit(dimnames(geno_obj)[[3]], "_"), function(x) x[2]))
pheno_obj$marker_location <- marker.pos

```

## Allele Frequency

```{r allele_freq, fig.width = 10, fig.height = 5}
allele_freq <- apply(geno_obj, 3, colMeans)
max.freq <- max(allele_freq)

chr.idx <- lapply(unique(pheno_obj$chromosome), function(x) which(pheno_obj$chromosome == x))
#quartz()
plot.new()
plot.window(xlim = c(0, ncol(allele_freq)), ylim = c(0, max(allele_freq)))
par(xpd = TRUE)
for(ch in 1:length(chr.idx)){
    text(x = mean(chr.idx[[ch]]), y = (max.freq*-0.05), labels = ch)
    if(ch %% 2 == 0){
      draw.rectangle(min(chr.idx[[ch]]), max(chr.idx[[ch]]), 0, max.freq, fill = "gray80",
      border = "gray80")
    }
  }
par(xpd = FALSE)
for(i in 1:nrow(allele_freq)){
  points(allele_freq[i,], type = "l", col = i)
}
axis(2)
mtext(side = 3, "Allele Frequencies")
```

Remove the first half of chromosome 9 because it has
allele frequencies hitting 0 and 1.

```{r chr9}
chr9.locale <- which(pheno_obj$chromosome == 9)
to.remove <- chr9.locale[1:(length(chr9.locale)/2)]

pheno_obj$chromosome <- pheno_obj$chromosome[-to.remove]
pheno_obj$marker_num <- pheno_obj$marker_num[-to.remove]
pheno_obj$marker_location <- pheno_obj$marker_location[-to.remove]
pheno_obj$geno_names[[3]] <- pheno_obj$geno_names[[3]][-to.remove]
saveRDS(pheno_obj, here("Results", "General", "pheno_obj.RDS"))

geno_obj <- geno_obj[,,-to.remove]
saveRDS(geno_obj, here("Results", "General", "geno_obj.RDS"))
```

## Trait Structure

The following correlation heat map shows that there are some pretty highly
correlated traits.

```{r trait_cor}
trait.cor <- cor(pheno_obj$pheno, use = "pairwise.complete.obs")
pheatmap(trait.cor, display_numbers = TRUE)
```

The following plot shows the first two PCs of the trait correlation matrix.
The traits split into two major groups along the 1st PC, and then two more
minor groups along the second. 

```{r trait_decomp}
trait.cor.decomp <- plot.decomp(trait.cor, label.points = TRUE)
```

```{r create_cape_obj}
#initialize a cape object and then run singlescan
singlescan.dir <- here("Results", "all_singlescan")
cape_obj_file <- file.path(singlescan.dir, "cross.RDS")

saveRDS(pheno_obj, file.path(singlescan.dir, "pheno_obj.RDS"))
if(!file.exists(cape_obj_file)){
  cape_obj <- run_cape(pheno_obj, geno_obj, 
    results_file = "cross.RDS", p_or_q = 0.05, n_cores = 4, 
    initialize_only = TRUE, verbose = TRUE, run_parallel = FALSE, 
    param_file = file.path(singlescan.dir, "cape.parameters.yml"), 
    yaml_params = NULL, results_path = singlescan.dir)
}else{
  cape_obj <- readRDS(cape_obj_file)
}

cape_geno <- readRDS(file.path(singlescan.dir, "cross_geno.RDS"))
```

```{r singlescan}
singlescan_obj_file <- file.path(singlescan.dir, "cross_singlescan.RDS")

if(!file.exists(singlescan_obj_file)){
  singlescan_obj <- singlescan(cape_obj, cape_geno, 
    kin_obj = NULL, n_perm = cape_obj$singlescan_perm, 
    alpha = cape_obj$alpha, verbose = TRUE, run_parallel = FALSE, 
    n_cores = 4, model_family = "gaussian", overwrite_alert = FALSE)
  saveRDS(singlescan_obj, singlescan_obj_file)
}else{
  singlescan_obj <- readRDS(singlescan_obj_file)
}
```


## Screen for interesting markers {.tabset .tabset-fade .tabset-pills}

We want to screen for individual markers that have large phenotypic
effects. We will use these markers as query alleles in query_cape.
Here we test all markers against the strain mean.

The following plots show the effect sizes from the single scans for 
each trait. This will help us identify pleiotropy for selection of 
trait combinations.

```{r standard_singlescan}
get_model_effects <- function(y, x){
  model <- lm(y~x)
  return(coef(model))
}

all.allele.effects.file <- here("Results", "General", "allele.effects.RDS")

if(!file.exists(all.allele.effects.file)){
  all.allele.effects <- vector(mode = "list", length = ncol(pheno_obj$pheno))
  names(all.allele.effects) <- colnames(pheno_obj$pheno)
  for(ph in 1:ncol(pheno_obj$pheno)){
    print(colnames(pheno_obj$pheno)[ph])
    allele.effects <- matrix(NA, nrow = ncol(geno_obj), ncol = dim(geno_obj)[[3]])
    for(a in 1:ncol(geno_obj)){
      print(a)
      allele.mat <- geno_obj[,a,]
      effects <- apply(allele.mat, 2, function(x) get_model_effects(pheno_obj$pheno[,ph], x))
      allele.effects[a,] <- effects[2,]
    }
    all.allele.effects[[ph]] <- allele.effects
  }
  saveRDS(all.allele.effects, all.allele.effects.file)
}else{
  all.allele.effects <- readRDS(all.allele.effects.file)
}

```

```{r plot_effects, results = "asis", fig.width = 10, fig.height = 6}
col <- categorical_pal(8)
u_chr <- unique(pheno_obj$chromosome)
chr_idx <- lapply(u_chr, function(x) which(pheno_obj$chromosome == x))
pos <- lapply(chr_idx, function(x) pheno_obj$marker_location[x])

#pdf("~/Desktop/yeast.pdf", width = 10, height = 6)
for(ph in 1:length(all.allele.effects)){
  cat("###", colnames(pheno_obj$pheno)[ph], "{.tabset .tabset-fade .tabset-pills}\n")

  ylim <- c(min(all.allele.effects[[ph]], na.rm = TRUE), 
          max(all.allele.effects[[ph]], na.rm = TRUE))

  for(ch in 1:length(u_chr)){
    cat("#### Chr", ch, "\n")
    layout.mat <- matrix(c(1:2), ncol = 2, byrow = TRUE)
    layout(layout.mat, widths = c(1,0.2))

    #ylim <- c(min(all.allele.effects[[ph]][,chr_idx[[ch]]], na.rm = TRUE), 
    #        max(all.allele.effects[[ph]][,chr_idx[[ch]]], na.rm = TRUE))
    xlim = c(min(pos[[ch]]), max(pos[[ch]]))
 
    par(mar = c(4,4,4,0))
    plot.new()
    plot.window(xlim = xlim, ylim = ylim)
    for(a in 1:nrow(all.allele.effects[[ph]])){
      points(x = pos[[ch]], y = all.allele.effects[[ph]][a,chr_idx[[ch]]],
      col = col[a], type = "l", lwd = 3)
    }
  axis(1);axis(2)
  mtext(paste("Position on Chr", ch), side = 1, line = 2.5)
  mtext("t statistic", side = 2, line = 2.5)
  mtext(paste("Chr", ch, names(all.allele.effects)[ph]), side = 3, line = 2.5, font = 2, cex = 1.5)
  abline(h = 0)

  #legend
  par(mar = c(4,0,4,1))
  plot.new()
  plot.window(xlim = c(0,1), ylim = c(0,1))
  legend(x = 0, y = 1, lty = 1, col = col, legend = 1:8,
      lwd = 3)
  cat("\n\n")
  }
cat("\n\n")
}
#dev.off()
```


## CAPE singlescan {.tabset .tabset-fade .tabset-pills}

These plots show the singlescans relative to allele 5

```{r plot_singlescan, fig.width = 10, fig.height = 5, results = "asis"}
par(mfrow = c(2,1))
for(i in 1:ncol(singlescan_obj$singlescan_t_stats)){
  cat("###", colnames(cape_obj$pheno)[i], "\n")
  #quartz(width = 11, height = 5)
  sub.result <- singlescan_obj
  sub.result$singlescan_effects <- sub.result$singlescan_effects[,i,,drop=FALSE]
  sub.result$singlescan_t_stats <- sub.result$singlescan_t_stats[,i,,drop=FALSE]
  sub.result$locus_score_scores <- sub.result$locus_score_scores[,i,drop=FALSE]
  plot_singlescan(cape_obj, sub.result, color_scheme = "other")
  plot_singlescan(cape_obj, sub.result, color_scheme = "other", standardized = FALSE)
  cat("\n\n")
}
```

## Singlescans Compared

The following plot shows the log of the the locus-based 
effect sizes aligned across the different traits. There 
appear to be several pleiotropic bands particularly on 
chromosomes 7 and 15. The traits are ordered by the 
maximum effect on chromosome 15.

Hu, Rapa, and Sorb have similar allele effects at a similar
location on Chr 15 (see detailed figures above). That could 
be an interesting combination of traits to look at.

CaCl2, YPD, and FLZ might have some pleiotropy on Chr 7.
SORB also has a touch of a peak in this region.

```{r multi_lod, fig.width = 11, fig.height = 5, warning = FALSE, message = FALSE}
#scan1.result <- t(apply(singlescan_obj$singlescan_t_stats, 1, function(x) apply(x, 1, max)))
scan1.result <- singlescan_obj$locus_score_scores
u_chr <- unique(cape_obj$chromosome)
map.char <- lapply(u_chr, 
  function(x) cape_obj$geno_names[[3]][which(cape_obj$chromosome == x)])
map <- lapply(map.char, function(x) as.numeric(substr(x, 4, nchar(x))))
names(map) <- u_chr
for(i in 1:length(map)){
  names(map[[i]]) <- paste0("chr", i, "_", map[[i]])
}
rownames(scan1.result) <- unlist(sapply(map, names))
#quartz(width = 11, height = 5)
#order by pleiotropic locus on Chr 15
chr15.locale <- grep("chr15", (rownames(scan1.result)))
trait.order <- order(apply(scan1.result[chr15.locale,], 2, max))

#jpeg("~/Desktop/multilod.jpg", width = 9, height = 4, units = "in", res = 300)
multilod.plot(log10(scan1.result[,trait.order]), map, chr.name.y = 0.2, 
  lod.thresh = log10(7))
#dev.off()

#pairs(scan1.result)
```

```{r select_query_marker, eval = FALSE}
#use this code to identify a marker and alleles to put into Query_CAPE.Rmd

trait.name = "FLZ"
trait.idx <- which(colnames(pheno_obj$pheno) == trait.name)
trait.effects <- all.allele.effects[[trait.idx]]
effect.max.pos <- sapply(1:length(chr_idx), function(x) apply(trait.effects[,chr_idx[[x]]], 1, function(y) pos[[x]][which.max(y)]))
effect.min.pos <- sapply(1:length(chr_idx), function(x) apply(trait.effects[,chr_idx[[x]]], 1, function(y) pos[[x]][which.min(y)]))
rownames(effect.max.pos) <- rownames(effect.min.pos) <- 1:8
colnames(effect.max.pos) <- colnames(effect.min.pos) <- paste0("Chr", 1:ncol(effect.min.pos))

#use the figures and the charts to pick the position of 
#the allele with the minimum or maximum effect.
#For example, strain 2 has a large susceptibility allele
#on Chr 1 as shown in the plots above. The position of this
#allele is in effect.min.pos row 2, column 1 (36048)
#Use this strain and position information to run a query
#pick a neutral reference allele (one other than the 
#query allele) to put into the cape.parameters file
#The default is allele 5



```